### PR TITLE
Added class templates to the language.

### LINKED ISSUE #964: Rock doesn't always resolve type template instances, leading to backend crashes
For example:  
```ooc
Foo: cover template <T> {
    fn: Func <T> (T)

    init: func@ (=fn)
}

foo: Foo<Int>
```  
Here, `Foo<Int>` is never resolved (both with master and my class_template branch), init is never resolved, leading to =fn never being resolved and not getting a type, which the backend tries to write blindly.  
The result, unsurprisingly, is a segfault.  

(For some reason, the exact same code with class instead of cover and func instead of func@ has its instance resolved on my class_template branch).  

It looks like we are relying on something wholeAgain-ing at the moment, which should not be the case.

### PR DESCRIPTION
**Pull request by shamanas**
They can be extended from, either partially (e.g. `Bar: class template <T> extends Foo<Int, T>`) or fully (e.g. `IntWrapper: class extends Foo <Int>`).  

In addition to adding type templates, this PR closes #964.  

See fasterthanlime/nagaqueen#29 for the (one line) grammar changes.

### CODE REVIEW

**Comment by shamanas** [comment]
Oops.  
Those failing tests passed at some point earlier in the commits.  
That will teach me to not test the final version on my local machine.  

Anyway, I'll be on it

**Comment by shamanas** [comment]
Actually, they pass on my machine >_>  
Well, that is weird af.

**Comment by shamanas** [comment]
Seriously, those tests run fine on my machine...  
Well, I'm giving up for today before I fry my brain.

File: source/rock/middle/TypeDecl.ooc [hunk 1]
```diff
@@ -196,7 +220,13 @@ TypeDecl: abstract class extends Declaration {
     getInterfaceDecls: func -> List<InterfaceImpl> { interfaceDecls }
 
     hasMeta?: func -> Bool {
-        !isMeta
+        if (debugCondition()) {
+            "hasMeta called, they want %s / %p back" printfln(toString(), template)
+        }
```
Comment (vendethiel) [inline source/rock/middle/TypeDecl.ooc]: Nice one :P

File: source/rock/middle/TypeDecl.ooc [hunk 2]
```diff
@@ -524,8 +554,197 @@ TypeDecl: abstract class extends Declaration {
 
     }
 
+
+    /**
+     * Generates the name (fingerprint) of a type template instance from a realized type.
```
Comment (vendethiel) [inline source/rock/middle/TypeDecl.ooc]: That's akin to name mangling, right?

File: source/rock/middle/TypeDecl.ooc (hunk 1 above)
Comment (shamanas) [inline source/rock/middle/TypeDecl.ooc]: All credit to @fasterthanlime, this is a copy pasted piece of code (from the cover template impl) and I'm not ashamed to say it :P

File: source/rock/middle/TypeDecl.ooc (hunk 2 above)
Comment (shamanas) [inline source/rock/middle/TypeDecl.ooc]: Yes, this is mangling at the ooc level.

**Comment by vendethiel** [comment]
Cool stuff :D good job

**Comment by shamanas** [comment]
Thank you :)  

Hopefully I will find out what is causing Travis to choke on those two tests swiftly tomorrow and have it merged.

**Comment by kirbyfan64** [comment]
What version of rock is Travis using for bootstrap? That might be an issue...

**Comment by shamanas** [comment]
Travis is just using the latest make rescue build.  
Anyways, it manages to build itself with c_rock, then build rock with rock, so if it gets up to the tests, it means the resulting rock is good.

**Comment by fasterthanlime** [comment]
pretty sure travis is using quick-rescue? careful

**Comment by shamanas** [comment]
Travis does:

```yml
  - make -s rescue
  - make clean self
  - make clean self
```

So it makes c_rock, c_rock makes rock (make -s rescue), rock makes rock (make clean self), rock makes rock (makes clean self).  
I don't think it could be a bootstrapping issue anyway since it is failing on tests that used to pass.

Anyway, I'm on it right now

**Comment by shamanas** [comment]
I can now reproduce the two errors on a fresh install on my linux distro, so I'm ready to go to town :)

**Comment by shamanas** [comment]
I'm merging this but I'm sure there will be tonnes of bugs related to templates to fix, so please report as many as you can if you play with them :)  

Things to note at the moment:  
- You always need to fully qualify your typeArgs to use type templates (generics first, templates second)
- There is no template inferring mechanism at all at the moment (no `Foo new(42) => T == Int` or stuff like that)  
- You should be able to define generic class templates and extend normally (emphasis on should)  
- This also closes #964, so you won't get weird crashes when not using an instance's fields  

My focus will now be on #895 which bugged me quite a bit while making this (and there are a couple of workarounds here).

**Pull request merged**